flows/definition/flow.go

```diff
@@ -95,12 +105,38 @@ func (f *flow) Validate(assets flows.SessionAssets, context *flows.ValidationCon
 		}
 		seenUUIDs[utils.UUID(node.UUID())] = true
 
-		if err := node.Validate(assets, context, f, seenUUIDs); err != nil {
+		if err := node.Validate(f, seenUUIDs); err != nil {
 			return errors.Wrapf(err, "validation failed for node[uuid=%s]", node.UUID())
 		}
 	}
 
+	// extract all dependencies (assets, contacts)
+	deps := newDependencies(f.ExtractDependencies())
+
+	// and validate that all assets are available in the session assets
+	missingAssets := make([]assets.Reference, 0)
+	deps.refresh(sa, func(r assets.Reference) { missingAssets = append(missingAssets, r) })
+
+	if len(missingAssets) > 0 {
+		depStrings := make([]string, len(missingAssets))
+		for d := range missingAssets {
+			depStrings[d] = missingAssets[d].String()
+		}
+		return errors.Errorf("missing dependencies: %s", strings.Join(depStrings, ","))
+	}
+
 	f.valid = true
+	f.dependencies = deps
+	f.resultNames = f.ExtractResultNames()
```

Comment:
I think this is threadsafe... it's possible that multiple threads validate this flow object at the same time, but it shouldn't matter